enhancements/hypershift/disaster-recovery/hcp-etcd-backup-api.md

```diff
@@ -0,0 +1,554 @@
+
+Contributes significantly to the reliability of the product by promoting seamless integration, automation, and standardization, ultimately providing users with a more dependable and user-friendly experience.
+
+The API will help with disaster recovery duties, like Hosted Cluster migration or Hosted Cluster recovery, both highly demanded features by Red Hat customers.
+
+### User Stories
+
+- As a customer I want to perform etcd backups of my Hosted Clusters in a concrete time window and schedule at the pace my business needs.
+- As a customer/engineer/service provider I want to have the ability to restore etcd cluster in a new hosted cluster.
+- As a developer I want to declaratively create etcd backups using and upload it to Amazon S3.
+- As a developer I want to declaratively create etcd backups using and upload it to Azure.
+- As a customer I want to perform etcd backups of my Hosted Clusters in a concrete time window and schedule in the pace my bussiness needs.
+- As a customer I want documentation about how to perform the backup and recovery procedures in my Hosted Cluster.
+
+### Goals
+
+- Provide a consumable API for Etcd backup that could be usable in a Hosted Cluster restoration scenario. These are the features included:
+  - Schedulable.
+  - Capability to select the right Time Zone.
+  - Retention policy based on size of the backup or number of backups.
+  - Local storage of the backup (PVCs).
+  - Remote storage of the backup in a service provider (S3, Azure...).
+  - Restoration of the backup in another Hosted Cluster using the already existent API.
+
+### Non-Goals
+
+- Provide a full backup/restoration solution for a hosted cluster.
+  - This is not covering the Volumes migrations
+
+
+### Pros and Cons
+#### Pros
+
+**Visibility:** Having an API could enhance visibility within the HostedCluster (HC) object, which serves as the primary interface for customers to manage backups. In the event of a backup failure, in addition to log entries, it could update the status.conditions within the HostedCluster object.
+
+**Integration:** Integration with the product is strengthened, meaning that adding new features becomes more straightforward. Employing tightly coupled methods could lead to poor practices.
+
+**Alerting:** The controller can notify users through various channels:
+
+- HC status.conditions
+- Log entries
+- Events
+- Alerting
+- Metrics
+
+**Validations:** In this scenario, we incorporate an additional layer of validations (CEL), managed by api-machinery. This layer promptly detects any discrepancies with the API consensus and triggers early failure. Furthermore, additional validations can be employed within the controller.
+
+**Reusability:** Leveraging the existing ETCD backup API developed by the ETCD team, we will integrate the entire API into a new structure within our ETCD API group. Within the same group, we will incorporate the necessary functionality to upload binaries to remote storage such as S3 or Azure.
```

Comment:
Does it always have to be remote storage? What would we do with agent, disconnected, kubevirt situations?

Reply:
Yes, the plan is to include some kind of block storage outside of this cluster (e.g: Minio or something similar) to be compatible with the S3 interface. This will cover the disconnected/bare metal providers.

Reply:
> within our ETCD API group

We don't have an etcd API group

Reply:
I would double-check with the field, but S3 is not very common on-prem.  Traditional on-prem is more likely to have NFS.

Reply:
The S3 will be implemented once the AWS case is covered so MinIO will be automatically supported. The plan is add new providers and NFS could be one of them but we need to prioritise the most used ones (ODF, NFS or w/e other) but I don't have enough data to say now which providers will be included in the initial approach. We can talk about this during the development and in the meantime capture this feedback.